src/parser/binary_parser.ml

```diff
@@ -1029,12 +1059,13 @@ let sections_iterate (input : Input.t) =
   in
 
   let* () =
-    match data_count_section with
-    | None -> Ok ()
-    | Some len ->
-      if List.compare_length_with data_section len <> 0 then
-        Error (`Msg "data count and data section have inconsistent lengths")
+    match (List.length data_section, data_count_section) with
+    | 0, None -> Ok ()
+    | _data_len, None ->
+      if code_use_dataidx then Error (`Msg "data count section required")
```

Comment:
It would be better to compute `code_use_data_idx` only in this case, to avoid traversing the whole code when it is not needed.